source/common/http/conn_manager_impl.h

```diff
@@ -439,8 +448,37 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
 
     std::chrono::milliseconds idle_timeout_ms_{};
     State state_;
+
     const bool expand_agnostic_stream_lifetime_;
+
+    // Snapshot of the route configuration at the time of request is started. This is used to ensure
+    // that the same route configuration is used throughout the lifetime of the request. This
+    // snapshot will be cleared when the cached route is blocked. Because after that we will not
+    // refresh the cached route and release this snapshot can help to release the memory when the
+    // route configuration is updated frequently and the request is long-lived.
+    Router::ConfigConstSharedPtr snapped_route_config_;
+    Router::ScopedConfigConstSharedPtr snapped_scoped_routes_config_;
+    // This is used to track the route that has been cached in the request. And we will keep this
+    // route alive until the request is finished.
     absl::optional<Router::RouteConstSharedPtr> cached_route_;
+    // This is used to track whether the route has been blocked. If the route is blocked, we can not
+    // clear it or refresh it.
+    bool route_cache_blocked_{false};
+    // This is used to track routes that have been cleared from the request. By this way, all the
+    // configurations that have been used in the processing of the request will be alive until the
+    // request is finished.
+    // For example, if a filter stored a per-route config in the decoding phase and may try to
+    // use it in the encoding phase, but the route is cleared and refreshed by another decoder
+    // filter, we must keep the per-route config alive to avoid use-after-free.
+    // Note that we assume that the number of routes that have been cleared is small. So we use
+    // inline vector to avoid heap allocation. If this assumption is wrong, we should consider using
+    // a list or other data structures.
+    //
+    // TODO(wbpcode): This is a helpless compromise. To avoid exposing the complexity of the route
+    // lifetime management to every HTTP filter, we do a hack here. But if every filter could manage
+    // the lifetime of the route config by itself easily, we could remove this hack.
```

Comment:
Just pointing out that if some filter does somehow use the extended route table after the response has been sent (logging callback, onDestroy, who knows), this is still going to crash. But I guess we can see?

Reply:
What do you mean by the `extended route table`? The `Router::ConfigImpl` instance? I think the filter should never and cann't access the whole route table? 🤔

Reply:
Sorry I just meant that it's possible for a filter to use route data after response headers have been sent (logging, onDestroy, etc.). But I guess now thinking about it, anything that a filter could access via raw pointer is kept alive via the hack, right? So there is no chance of crash in any further use?